system/Validation/FormatRules.php

```diff
@@ -291,6 +294,32 @@ public function valid_url(?string $str = null): bool
         return filter_var($str, FILTER_VALIDATE_URL) !== false;
     }
 
+    /**
+     * Checks a URL to ensure it's formed correctly.
+     *
+     * @param string|null $validSchemes comma separated list of allowed schemes
+     */
+    public function valid_url_strict(?string $str = null, ?string $validSchemes = null): bool
+    {
+        if (empty($str)) {
+            return false;
+        }
+
+        if ($validSchemes === null) {
+            $validSchemes = 'http,https,mailto,tel,sms';
+        }
+
+        $scheme           = strtolower(parse_url($str, PHP_URL_SCHEME));
+        $validSchemes     = strtolower($validSchemes);
+        $validSchemeArray = explode(',', $validSchemes);
+
+        if (! in_array(($scheme), $validSchemeArray, true)) {
+            return false;
+        }
+
+        return filter_var($str, FILTER_VALIDATE_URL) !== false;
```

Comment:
Maybe something like this?
```suggestion
        $scheme       = strtolower(parse_url($str, PHP_URL_SCHEME));
        $validSchemes = explode(',', strtolower($validSchemes ?? 'http,https,mailto,tel,sms'));

        return in_array($scheme, $validSchemes, true)
                ?: filter_var($str, FILTER_VALIDATE_URL) !== false;
        }
```

Reply:
A bit different. I pushed the refactoring.

Reply:
The return can be simplified further to use &&.

return in_array() && filter_var();

Reply:
Refactored.